### PR TITLE
Fix transfer selected keys

### LINKED ISSUE #51497: Transfer 穿梭框组件 同时更新 dataSource和selectedKeys 会导致无法初始化选中状态
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/p/sandbox/ji-ben-yong-fa-antd-5-21-6-forked-dh7h8k?workspaceId=bfe5e21e-2e78-4945-8842-a8e96aafb52e)

### Steps to reproduce

打开code链接，点击update，应该选中的条目没有被选中

### What is expected?

按照selectedKeys给出的key值选中条目

### What is actually happening?

列表数据更新了，但是勾选状态没有更新

| Environment | Info |
| --- | --- |
| antd | 5.21.6 |
| React | 18.0.0 |
| System | all |
| Browser | all |

### PR DESCRIPTION
**Pull request by IsKaros**
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE_CN.md?plain=1)

### 🤔 This is a ...

- [ ] 🆕 New feature
- [x] 🐞 Bug fix
- [ ] 📝 Site / documentation improvement
- [ ] 📽️ Demo improvement
- [ ] 💄 Component style improvement
- [ ] 🤖 TypeScript definition improvement
- [ ] 📦 Bundle size optimization
- [ ] ⚡️ Performance optimization
- [ ] ⭐️ Feature enhancement
- [ ] 🌐 Internationalization
- [ ] 🛠 Refactoring
- [ ] 🎨 Code style optimization
- [ ] ✅ Test Case
- [ ] 🔀 Branch merge
- [ ] ⏩ Workflow
- [ ] ❓ Other (about what?)

### 🔗 Related Issues

- fix #51497

### 💡 Background and Solution
fix `selectedKeys` after changing dataSource by taking prop `selectedKeys`  into  consideration 

### 📝 Change Log

fix `selectedKeys` after changing dataSource

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   fix `selectedKeys` after changing dataSource        |
| 🇨🇳 Chinese |    修复在变更数据源之后`selectedKeys` 异常的bug       |

### CODE REVIEW

**Comment by yoyo837** [comment]
Could you add a test case for this?

File: components/transfer/__tests__/index.test.tsx [hunk 1]
```diff
@@ -767,6 +768,71 @@ describe('Transfer', () => {
...
+
+    defaultCheckedKeys.forEach((item) => {
+      expect(
+        container
+          ?.querySelectorAll('.ant-transfer-list-content-item')
+          ?.item(Number(item))
+          ?.querySelector('input[type="checkbox"]')!,
+      ).toBeChecked();
+    });
+    // expect(handleSelectChange).toHaveBeenLastCalledWith(defaultCheckedKeys, []);
```
Comment (zombieJ) [inline components/transfer/__tests__/index.test.tsx:834]: remove useless code

File: components/transfer/hooks/useSelection.ts [hunk 2]
```diff
@@ -48,8 +48,10 @@ export default function useSelection<T extends { key: TransferKey }>(
 
   // Reset when data changed
   React.useEffect(() => {
-    setSourceSelectedKeys(filterKeys(sourceSelectedKeys, leftKeys));
-    setTargetSelectedKeys(filterKeys(targetSelectedKeys, rightKeys));
+    const sourceKeysToSelect = selectedKeys.length > 0 ? selectedKeys : sourceSelectedKeys;
```
Comment (zombieJ) [inline components/transfer/hooks/useSelection.ts:51]: 过去这个实现感觉有点怪异，现在看起来似乎直接 useMemo 派生出来就行了。不需要内部有个 state 还通过 effect 来更新。

Comment (IsKaros) [reply components/transfer/hooks/useSelection.ts:51]: 我的想法是，当有selectedKeys时以selectedKeys为基础进行filter，当没有selectedKeys时，以上一次的结果为准进行filter。只用useMemo好像不能拿到上一次的结果。

Comment (zombieJ) [reply components/transfer/hooks/useSelection.ts:51]: `useMemo` 不用拿上次结果，每次都是计算数据。一定是最新的~

**Review by zombieJ — APPROVED** [review]

**Pull request merged**